### DISCUSSION TITLE
Any call for GetRecords giving no records with an error message.

### DISCUSSION THREAD

**Discussion by PrivateDave** [JoshClose/CsvHelper#1253] [Q&A]
In the case of that error, your library is for me totally useless. Do you have an idea of what is going on?

The exception occurs as a return error value of GetRecord function:

Evaluation of method System.Linq.SystemCore_EnumerableDebugView`1[<>f__AnonymousType0`2[System.Int64,System.String]].get_Items() calls into native method Microsoft.Win32.Win32Native.ReadFile(Microsoft.Win32.SafeHandles.SafeFileHandle, byte*, int, int&, System.IntPtr). Evaluation of native methods in this context is not supported.

and the stack trace: 
```
   at CsvHelper.CsvParser.Read()
   at CsvHelper.CsvReader.Read()
   at CsvHelper.CsvReader.<GetRecords>d__63`1[[<>f__AnonymousType0`2[[System.Int64, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.String, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], CSVHelperCoreTest, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].MoveNext()
   at System.Collections.Generic.LargeArrayBuilder`1[[<>f__AnonymousType0`2[[System.Int64, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.String, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], CSVHelperCoreTest, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].AddRange(IEnumerable`1 items)
   at System.Collections.Generic.EnumerableHelpers.ToArray[<>f__AnonymousType0`2](IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[<>f__AnonymousType0`2](IEnumerable`1 source)
   at System.Linq.SystemCore_EnumerableDebugView`1[[<>f__AnonymousType0`2[[System.Int64, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.String, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], CSVHelperCoreTest, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].get_Items()
```

**Comment by PrivateDave** [comment]
For sure problem lay in CsvReader which is simply empty and not initialized with data from stream reader.

**Comment by JoshClose** [comment]
Are you able to supply sample code that reproduces the issue?

**Comment by MunkeeJuice** [comment]
I am seeing exactly the same issue.  The project is .Net Core 2.2

>             using (TextReader reader = new StreamReader("test.csv"))
        {
                var csvReader = new CsvHelper.CsvReader(reader);
                var records = csvReader.GetRecords<dynamic>();
            }

The csv file has two cols Id and Word.  I have tried using a class which has Id and Word as public properties but I am still seeing the same issue.  Nothing fancy about it just using this code to test the library.

**Comment by JoshClose** [comment]
@MunkeeJuice Can you create sample code or a project, including the data, that reproduces this?

**Comment by MunkeeJuice** [comment]
> @MunkeeJuice Can you create sample code or a project, including the data, that reproduces this?

Sure can, do you just want me to add a sample to github and link you to the repo?  Wont be able to do this until tomorrow i'm afraid, but i will do it first thing in the morning when i get to the office.

**Comment by MunkeeJuice** [comment]
I have have made a test app at home but this was built under VS 2017 (getting a totally different error on that one) not 2019 like the one at work where i am seeing the above issue.  I will create a test project in 2019 in the morning and link you to the repo once it's done.

**Comment by MunkeeJuice** [comment]
Ok i rebuilt the project that i did at home in VS 2017 at work this morning in VS 2019

and am now getting this error 

Evaluation of method System.Linq.SystemCore_EnumerableDebugView`1[CsvHelper_Test.Test].get_Items() calls into native method Interop+Kernel32.ReadFile(System.Runtime.InteropServices.SafeHandle, byte*, int, int&, System.IntPtr). Evaluation of native methods

The test project can be found here https://github.com/MunkeeJuice/CsvHelper_Test

**Comment by VonPittman** [comment]
I had this exact issue and was able to find a solution:

First I had to turn on ProjectName > Properties > Debug > Enable native code debugging. Don't know why this was off by default, but it was. 

Then a more reasonable error message popup showed me the MissingFieldException error, and it instructed me to "Ignore missing fields by settings MissingFieldFound to null. 

This was accomplished with: `cr.Configuration.MissingFieldFound = null`
where cr is my CsvReader.

Hope this helps.

**Comment by abbottmw** [comment]
I encountered this issue, and I was unable to get it to work calling GetRecords<T>().  It works if i loop through the results using a read() and GetRecord()

I made a netfiddle.

[Calling GetRecords<Foo>() generates error](https://dotnetfiddle.net/upzVLw)

**Comment by AltruCoder** [comment]
@abbottmw Your fiddle doesn't generate an error for me.  `csv.GetRecords<Test>();` doesn't produce any results because you have already read to the end of the `TextReader` with `while (csv.Read())`.  If you remove the `while (csv.Read())` block then `var records = csv.GetRecords<Test>();` will work as expected.
```csharp
public class Program
{
	public static void Main()
	{
		string foo = @"1,one";
		
		using (TextReader reader = new StringReader(foo))
		using (var csv = new CsvReader(reader))
		{   
			csv.Configuration.HasHeaderRecord = false;			
			
			//this works
			var records = csv.GetRecords<Test>();
			Console.WriteLine(JsonConvert.SerializeObject(records));			
		}		
	}
}

public class Test
{
	[Index(0)]
	public int Id { get; set; }
	[Index(1)]
        public string Name { get; set; }
}
```

**Comment by abbottmw** [comment]
Okay that worked for me thank you, but what still isnt working is I have a DataParser class i wrote that contains all the csvHelper calls and a Parse() method that returns the csv.GetRecords(), but I cannot do anything with it as it has an error.  Probably user error.

What am I doing wrong here?

[updated fiddle](https://dotnetfiddle.net/isJHj1)

**Comment by AltruCoder** [comment]
`GetRecords<T>()` doesn't get the records right away.  It `yields` the results as you iterate the records.  Since you return the `IEnumerable` outside of the `using (var csv = new CsvReader(reader))`, you no longer have access to `StringReader(foo)`.  You can fix this by enumerating all of the records with `.ToList()` before it leaves the using statement. 
```csharp
return csv.GetRecords<T>().ToList();
```
[CsvHelper Getting Started](https://joshclose.github.io/CsvHelper/getting-started/)
>The GetRecords<T> method will return an IEnumerable<T> that will yield records. What this means is that only a single record is returned at a time as you iterate the records. That also means that only a small portion of the file is read into memory.

**Comment by Shashikanthmeda** [comment]
> `GetRecords<T>()` doesn't get the records right away. It `yields` the results as you iterate the records. Since you return the `IEnumerable` outside of the `using (var csv = new CsvReader(reader))`, you no longer have access to `StringReader(foo)`. You can fix this by enumerating all of the records with `.ToList()` before it leaves the using statement.
> 
> ```cs
> return csv.GetRecords<T>().ToList();
> ```
> 
> [CsvHelper Getting Started](https://joshclose.github.io/CsvHelper/getting-started/)
> 
> > The GetRecords method will return an IEnumerable that will yield records. What this means is that only a single record is returned at a time as you iterate the records. That also means that only a small portion of the file is read into memory.

Worked for me. Thank you very much. :)
